hopeit-git/hopeit.engine
# feature(engine): Apps Client

MOTIVATION
==========
Microservices call each other and users usually create ad-hoc http calls for this, what it's fine and allowed, but in the context of hopeit.engine it could be better to standardize the way apps are calling each other allowing:

- [x] Keep track of app dependencies at configuration time
- [x] Make it easy for an app developer to invoke other apps
- [x] Monitor invokations and failures
- [x] Automatically implement retry and circuit breaking
- [x] Allowing load-balancing on client side
- [x] Integrate authentication

SOLUTION
=========
This PR aims to solve at least partially the above needs implementing:
- [x] a way to specify dependencies (app-connections) on app config files (see client-example)
- [x] basic client load-balancing
- [x] log calls with caller information when invoking external endpoints
- [x] automatic retry on failure
- [x] basic circuit breaker
- [x] being able to propagate or obtain authentication tokens through multiple apps

PR PROGRESS
============
- [x] App and event connection spec supported in AppConfig
- [x] Engine support to instantitate Clients as plugins
- [x] hopeit.apps_client plugin based on aiohttp
- [x] client_example application connecting to simple_example
- [x] apps_visualizer support to show event connections
- [x] support for app-to-app Bearer auth using JWT tokens
- [x] client example integration tests
- [x] bump version to 0.9.0, prepare release

REVIEW & TEST
=============

```
make install-examples

make PORT=8020 run-simple-example
make PORT=8021 run-simple-example
make PORT=8030 HOSTS=http://localhost:8020,http://localhost:8021 run-client-example
make PORT=8098 HOSTS=http://localhost:8020,http://localhost:8021,http://localhost:8030,http://localhost:8098,http://localhost:8099 run-apps-visualizer
make PORT=8099 run-log-streamer
```

Check:
- in localhost:8030/api/docs:
  - Login basic auth
  - Hit count_and_save endpoint, it should call simple_example endpoints (it will create a Bearer token for each call). This works cause public key of client_example app is in .secrets/public/ where simple-example runs.
  - Kill one simple-example service and check that endpoints still serves using retries
  - Kill both and check NoHostsAvailable error coming back
  - restart simple-example, and check after 60 seconds that circuit breaker closes

- Open apps_visualizer (localhost:8098/api/docs) and check that client-examples is connected to simple-example
